### PR TITLE
Expose Swift Mailer streaming options in config, fixes #12702

### LINKED ISSUE #12702: Expose Mailer stream options (setStreamOptions) in configuration
**tl;dr** Filing a request for new feature that puts content from configuration (`config.php`) to `setStreamOptions`. _I am also available to create/send patch_, but wanted to first file the issue and see if this would get be accepted. Since `setStreamOptions` accept ordinary array, it could fit nicely in `config.php` IMHO, as `mail_smtpstreamoptions` (following current namespaces there)

### Steps to reproduce

I am having Nextcloud locally (QNAP), but mail provider that I want to use is fastmail.

* Fastmail is fairly advanced mail provider, yet I didn't found that they have ability to allow less secure apps, like [google](https://support.google.com/accounts/answer/6010255?hl=en) is having.
* I am using SMTP as an `mail_smtpmode` and none of the configuration parameters is working for me.
* I also cannot change QNAP easily (add cert/move instance domain)

So, at the end, I have to change Nextcloud:) So, based on [this](https://stackoverflow.com/questions/36454920/swiftmailer-php-how-to-disable-ssl-certificate-validation) this is my dance I am doing whenever Mailer.php is overwitten:
```php
$transport->setStreamOptions(array('ssl' => array('allow_self_signed' => true, 'verify_peer' => false, 'verify_peer_name' => false)));
```
(in `Mailer.php` in `getSmtpInstance`)

We might add separate config param to allow specifically this (allow_self_signed, _verify_peer=false...), but this is just promoting insecure usage and exposing `setStreamOptions` seems like a more generic solution.

### Expected behaviour

Test mail should be sent when SMTP with my setup is sent

### Actual behaviour

```
Uncaught Swift_TransportException: Connection could not be established with host smtp.fastmail.com
```

### Server configuration

**Operating system**:

QNAP

**Web server:**

Apache2

**Database:**

PGSQL

**PHP version:**

7.x

**Nextcloud version:** (see Nextcloud admin page)

14.x

**Updated from an older Nextcloud/ownCloud or fresh install:**

Updated

**Where did you install Nextcloud from:**

tar.gz

### PR DESCRIPTION
**Pull request by stalker314314**
Backward compatible addition to expose streaming options, so user can define them in their repsective configs.

Swift Mailer added streaming options 3 years ago in [this commit](https://github.com/swiftmailer/swiftmailer/commit/d791ebe2266d5f997a6f712fa8892620395b0c0e). Seems Nextcloud is having proper version.

Fixes #12702

### CODE REVIEW

**Comment by stalker314314** [comment]
Friendly ping:)

**Comment by ChristophWurst** [comment]
Hey @stalker314314,

thanks for your contribution :v: 

I'm sorry this got no attention yet, we release Nextcloud 15 today and we were busy preparing that.

Let me assign the corresponding labels so this gets reviewed and merged soonish :)

**Comment by stalker314314** [comment]
Ping. Pending for quite some time now! Can someone please take a look?

File: lib/private/Mail/Mailer.php [hunk 1]
```diff
@@ -259,6 +259,10 @@ protected function getSmtpInstance(): \Swift_SmtpTransport {
 		if (!empty($smtpSecurity)) {
 			$transport->setEncryption($smtpSecurity);
 		}
+		$streamingOptions = $this->config->getSystemValue('mail_smtpstreamoptions', array());
+		if (is_array($streamingOptions) && count($streamingOptions) > 0) {
```
Comment (nickvergessen) [inline lib/private/Mail/Mailer.php]: `!empty($streamingOptions)` instead of the count

**Comment by nickvergessen** [comment]
The only problem I have with this is that your config is then specific to Swift Mailer.
But well, if we ever replace it we will find a way.

**Review by nickvergessen — APPROVED** [review]

File: lib/private/Mail/Mailer.php (hunk 1 above)
Comment (stalker314314) [reply lib/private/Mail/Mailer.php]: done

File: lib/private/Mail/Mailer.php [hunk 2]
```diff
@@ -259,6 +259,10 @@ protected function getSmtpInstance(): \Swift_SmtpTransport {
 		if (!empty($smtpSecurity)) {
 			$transport->setEncryption($smtpSecurity);
 		}
+		$streamingOptions = $this->config->getSystemValue('mail_smtpstreamoptions', array());
```
Comment (skjnldsv) [inline lib/private/Mail/Mailer.php]: Please use [] not array() more readable

File: tests/lib/Mail/MailerTest.php [hunk 3]
```diff
@@ -167,4 +167,26 @@ public function testCreateEMailTemplate() {
 
 		$this->assertSame(EMailTemplate::class, get_class($this->mailer->createEMailTemplate('tests.MailerTest')));
 	}
+
+	public function testStreamingOptions() {
+		$this->config->method('getSystemValue')
+			->will($this->returnValueMap([
+				['mail_smtpmode', 'smtp', 'smtp'],
+				['mail_smtpstreamoptions', array(), array('foo' => 1)]
```
Comment (skjnldsv) [inline tests/lib/Mail/MailerTest.php]: Same here, don't mix `array()` and `[]`

File: lib/private/Mail/Mailer.php (hunk 2 above)
Comment (stalker314314) [reply lib/private/Mail/Mailer.php]: fixed

File: tests/lib/Mail/MailerTest.php (hunk 3 above)
Comment (stalker314314) [reply tests/lib/Mail/MailerTest.php]: fixed

**Review by skjnldsv — APPROVED** [review]
:+1: :tada:

**Comment by skjnldsv** [comment]
Failure unrelated.

**Comment by skjnldsv** [comment]
@stalker314314 Could you add your config to the documentation please? :)

**Pull request merged**